docs/refman/ppr-lp.1.pod
#
# mouse:~ppr/src/docs/refman/ppr-lp.1.pod
# Copyright 1995--2005, Trinity College Computing Center.
# Written by David Chappell.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# 
# * Redistributions of source code must retain the above copyright notice,
# this list of conditions and the following disclaimer.
# 
# * Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
#
# Last modified 14 January 2005.
#

=head1 NAME

ppr-lp - selects spooler and submits print job

=head1 SYNOPSIS

B<ppr-lp> I<switches> [I<filename ...>]

=head1 DESCRIPTION

This program is part of the PPR uprint suite.

Your system administrator may have chosen to replace B<lp> with a link to
this program.

This program is designed to be a drop-in replacement for the B<lp> program from
System V Unix.  As such it follows B<lp> sematics as much as possible in matters such
as deciding when to send e-mail to the job submitter, and determining the queue to 
which the job is to be sent.

=over 4

=item B<-d> I<queue>

This option specifies the print queue in which the job should be entered.
If this option is omited, then the LPDEST and PRINTER environment variables are
searched in that order.  If they are both undefined, then a queue name
is taken from the "ppr-lp=" setting in the "[default destinations] section
of F<uprint.conf>.  If that too is not defined, then a queue called "lp" is
assumed.

Notice that it is quite possible that B<ppr-lp> with the B<-d> option will
select a different queue from what B<ppr> with the B<-d> option or B<ppr-lpr>
without the B<-P> option would choose.

=item B<-i> I<requestid>

This switch is used to specify the request id of a job to be manipulated
as discribed by the B<-?> switch.  This switch is not yet supported by
B<ppr-lp>.

=item B<-H> I<handling>

This switch requests "special handling".  The valid values for I<handling> are
"hold", "resume", and "immediate".  This switch does not work yet
in B<ppr-lp>.

=item B<-c>

This switch indicates that the spooler should make its own copy of the job
rather than reading the origional when it is ready to print.  This option
should be used if you will delete or modify the origional before printing
is likely to be finished.

=item B<-f> I<formname>

This switch specifies a "form" (generally a kind of paper) on which the
job should be printed.  Naturally, this will work with B<lp> queues.
For PPR queues, the values is considered to be a PPR medium name fed
to B<ppr> using its B<-D> (default medium) switch.  For B<lpr> queues
this switch will have no effect.

=item B<-m>

Send email when the job has been printed.

=item B<-w>

Write a message to the user's terminal when the job is done.

=item B<-n> I<n>

Print the indicated number of copies.

=item B<-o> I<options>

This sets "interface options".  The options are expressed as name=value pairs.
The standard interface options were conceived in the days when line printers
dominated.  It is often difficult to determine how they ought to apply to modern
printers such as laser printers.

The standard interface options are as follows:

=over 4

=item nobanner

=item nofilebreak

=item length=

=item width=

=item lpi=

=item cpi=

=item stty=

=back

=item B<-y> I<list>

This switch gives a list of "modes".  These are operations performed on
the file by filters.  The list of available modes is defined by the
system administrator.  The effect of this switch on non-B<lp> queue is
not yet defined.

=item B<-P> I<list>

This option can be used to indicate that only certain page should be printed.
The real B<lp> man page says that requests of this type can only be
satisfied when a filter exists that can extract pages and that the job
will be rejected if the request cannot be satisfied.  This option is presently
ignored when the queue is a B<lpr> queue.  PPR does not yet implement page
selection and so will ignore such a request.

=item B<-p>

This option "enables notification", according to the real B<lp> man page.
It is unclear what this means.

=item B<-q> I<n>

This sets the queue priority of the job.  Queue priorities range from 0
(the highest) to 39 (the lowest).  This parameter has not effect on
B<lpr> queues but it works with B<lp> and PPR queues.

=item B<-s>

This option suppresses the display of certain messages.  Precisely what
messages ought to be suppressed is unclear but the message which indicates
the request id of the newly created job is among them.

=item B<-S> I<charset>

Specify the character set.  For the B<lp> spooler it appears that the
list of character set names is defined by the system administrator, though
there may be some standard names.  This option is ignored by B<lpr> queues.
For PPR queues, the I<charset> is passed using PPR's "charset=" filter option.
Acceptable character set names for PPR include "latin1" and "cp437".

=item B<-t> I<string>

This switch is used to specify a job name (title) for the banner page.

=item B<-T> I<type>

This switch indicates the format (type) of the input file.

For B<lp> queues, these types are defined by the system administrator,
however there are a few
common type name including "simple" and "postscript".

For B<lpr> queues, a few common type names are converted to the appropriate
B<lpr> switch.  The translation mechanism is in flux and is not documented
at this time.

For PPR queues, the indicated I<type> is passed to B<ppr>'s B<-T> switch.

=item B<-r>

This indicates that the input file should be passed through to the printer
without filtering.  The precise effect of this switch on non-B<lp> queues
is not yet defined.

=back

=head1 ENVIRONMENT VARIABLES

To learn how the environment variables LPDEST and PRINTER are used, see the
description of the B<-d> option.

The variables PPR_RESPONDER, PPR_RESPONDER_ADDRESS and PPR_RESPONDER_OPTIONS
can be used to control the response method used to inform the submitter of 
the fate of such jobs as are eventually processed by PPR (as opposed to some 
other spooler.  They work as described in the L<ppr.1> manpage.  If the job is
passed to a remove node's PPR spooler through the LPR/LPD protocol and 
B<lprsrv>, a protocol extension is used to convey this information.


=head1 BUGS

While B<ppr-lp> is intended to be a drop-in replacement for B<lp>, not all
possible options and option values are implemented.  Sometimes this is because
the features in question simply aren't supported by B<ppr> or don't apply to 
PostScript printers.  However, others are not implemented or may not be
implemented correctly due to a lack of information.  Bug reports are welcome 
and should be sent to ppr-bugs@trincoll.edu.

=head1 AUTHOR

David Chappell, Trinity College Computing Center, Hartford, Connecticut.

=cut



History.txt
mouse:~ppr/src/History.txt (symbolic link)

This file contains a history of changes to PPR which occured after the
release of version 1.00.


1.00, 26 April 1995

First public release.


1.10, 27 April 1995--7 June 1995

Removed "From:" line from message passed to Sendmail by pprd when sending
a message listing the remaining printer faults.  This is so that Sendmail
will take the "real name" from the password file and put it in the "From:"
line.

Removed dotmatrix/input.epson, a GIF file which had been inadvertently
included in the distribution.

Moved the contents of include/atalk_streams.h into
interfaces/atalk_streams.c since it was not long enough to justify an
include file, nor was it of interest to any other program.

Added code to ppad to classify members of the group "ppradmin" as PPR
addministrators who can edit printer and group configurations.

Added -v switch to ppad.

Removed uid's 1-99 from the list of priveledged users.

Fixed ppad(8) so that it considers the group "ppradmin" as priveledged
rather than "ppr".

Changed ppr(1) so that it will accept the members of the group "pprprox" as
priveledged as well as the members of "ppr".  This is for security reasons
as one may not want to grant a user who has the privedge of using the -f
switch the priveledge of deleting other people's jobs.

Modified ppop(1) and pprd(8) so that user's can request to see the queue
entry for a single job.  Previously, the only options were all jobs for a
certain destination or all jobs in all queues.

Rewrote and renamed the undocumented "ppop qdebug" command to become the new
documented command "ppop details".  (This was done primarily to support
wwwppop, an experimental managment interface for PPR.  Since it was still
incomplete it was not released with this version of PPR.)

Renamed interfaces/atalk_streams.c to atalk_ali.c and
papsrv/papsrv_streams.c to papsrv_ali.c

Removed code from "make dirs" so that it doesn't create the directory
"/etc/ppr/groups.ppd".

Added sleep=N and eoftimeout=N options to tcpip interface.

Added -C _filename_ switch to ppr invokation in xppr.

Added "-default okay" to xmessage invokation in xppr.

Added the "ppop lpq" sub-command to the "ppop help" information, thus
making it a documented sub-command.

Extended automatic bin selection to multiple media types when "%%PageMedia:"
comments are present.  This has not yet been well tested.

Made "%%EndProcset" an acceptable substitute for "%%EndProcSet" due to a
reputed bug in QuickDraw GX.

Changed rule in ppr_infile.c for detecting PCL code so that documents with
only one escape would still be detected.


1.20, 8 June 1995--10 January 1996

Fixed code in ppr_infile.c so that the ppr -o switch will work even if there
is no "DefFiltOpts:" line in the printer or group configuration file.

Added '(' to list of likely second characters for HP escape codes
in ppr_infile.c

Replaced printing of errno value with strerror(errno) output in some places,
eliminated printing of errno=x from other places where both were printed.

Fixed ppr(1) so that it sets the environment variables "PATH", "IFS", and
"SHELL" so that filters and responders it invokes are less easy to subvert.

Removed setting of PATH from ~ppr/src/misc_filters/tex.perl and changed
shell name from "/usr/bin/tainperl" to "/usr/bin/perl".

Changed code which determins default responder address so that the
environment variables "USER" and "LOGNAME" are used in preference to the
current uid.

Added prompts in Configure to set the directory into which the links to PPR
binaries will be written and the path to the X-Windows binaries such as
xterm.

Modified papsrv so that it will pass the network and node number of the
client to ppr with the -r switch.  (This is for use with an experimental
AppleTalk responder.)

Modified papsrv.c, renaming cleanup() to termination_handler() and causing
it to call fatal() only the first time it is called.  Without this, name
could be left on the network if the CAP version received two terminate
signals in rapid sucession.

Added code to re-line-wrap responder messages for certain responders.
(The library function responder_width() returns the width for a responder.
Currently it is hard coded to return zero for all responders except "samba".
A return value of zero indicates that the message should not be
re-line-wrapped.)

Dropt reading of LPDEST and PRINTER by ppr(1).  (I can not really remember
why I put this feature in.  I don't think the reasons where very good.  It
may have been that I thought that the same printer would be available thru
two spooling systems so one might want to set the default destination for
both with one variable.  That now seems a little far fetched.)

Separated Solaris and Starserver sections of sysdep.h

Added PNG (Portable Network Graphics) format with magic number recognition
to ppr_infile.c.  This remains untested since I do not have any programs
which manipulate this graphics format.

In pprd.c, put sleep(1) between execl("pprdrv") and _exit(242) so that pprd
has time to record the PID before the child exits.  (If this is not done
and execl() fails immediately, a child which should have become pprdrv may
exit before the parent records its PID in the information structure for that
queue.  This result is that it will not recoginize the purpose of the child
when it receives notification of its death and will leave the queue in the
printing state indefinitely.)

Did same in pprdrv.c in code which executes interface.  Also changed
_exit(255) to _exit(242).  (Exit 242 is the new standard since PERL uses 255
as the exit code when die() is executed.)

Added automatic creation of "DefFiltOpts:" lines to ppad(8).  Manual
creation is supported by the the "ppad deffiltopts" and "ppad group
deffiltopts" command.

Modified ppop so that interactive mode restores the terminal settings when
it exits.  Also changed terminal setting code to use the POSIX method of

ppr-sysv/ppr-lp.c
/*
** mouse:~ppr/src/ppr-sysv/ppr-lp.c
** Copyright 1995--2005, Trinity College Computing Center.
** Written by David Chappell.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**
** * Redistributions of source code must retain the above copyright notice,
** this list of conditions and the following disclaimer.
** 
** * Redistributions in binary form must reproduce the above copyright
** notice, this list of conditions and the following disclaimer in the
** documentation and/or other materials provided with the distribution.
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
** AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
** IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
** ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
** LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
** CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
** SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
** INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
** CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
** ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
** POSSIBILITY OF SUCH DAMAGE.
**
** Last modified 14 January 2005.
*/

#include "config.h"
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <pwd.h>
#include <errno.h>
#ifdef INTERNATIONAL
#include <locale.h>
#include <libintl.h>
#endif
#include "gu.h"
#include "global_defines.h"
#include "uprint.h"

extern char *optarg;
extern int optind;

static const char *const myname = "ppr-lp";

void uprint_error_callback(const char *format, ...)
	{
	va_list va;
	fprintf(stderr, "%s: ", myname);
	va_start(va, format);
	vfprintf(stderr, format, va);
	fputc('\n', stderr);
	va_end(va);
	} /* end of uprint_error_callback() */

static const char *option_list = "cd:f:H:mn:o:P:pq:st:T:rwy:";

int main(int argc, char *argv[])
	{
	void *upr;
	struct passwd *pw;
	int c;
	const char *opt_i = (char*)NULL;	/* -i switch */
	const char *opt_H = (char*)NULL;	/* -H switch */

	/* Initialize internation messages library. */
	#ifdef INTERNATIONAL
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	#endif

	/*
	** Create a uprint object.  Notice that we pass the argument
	** list to the constructor.  If uprint_print() executes
	** /usr/bin/lp, it will pass it the origional arguments
	** rather than an reconstructed set.
	*/
	if((upr = uprint_new("lp", argc, (const char **)argv)) == (void*)NULL)
		{
		fprintf(stderr, "%s: %s\n", myname, uprint_strerror(uprint_errno));
		return 1;
		}

	/*
	** Set some defaults:
	*/
	uprint_set_show_jobid(upr, TRUE);
	uprint_set_immediate_copy(upr, FALSE);

	/*
	** Set the ID number and name of the user:
	*/
	{
	uid_t uid = getuid();

	if((pw = getpwuid(uid)) == (struct passwd *)NULL)
		{
		fprintf(stderr, _("%s: getpwuid(%ld) failed to find your account\n"), myname, (long int)uid);
		uprint_delete(upr);
		return 1;
		}
	if(pw->pw_name == (char*)NULL)
		{
		fprintf(stderr, "%s: strange getpwuid() error, pw_name is NULL\n", myname);
		uprint_delete(upr);
		return 1;
		}
	uprint_set_user(upr, uid, -1, pw->pw_name);
	}

	/*
	** Parse the switches.  Mostly, we will call uprint
	** member functions.
	*/
	while((c = getopt(argc, argv, option_list)) != -1)
		{
		switch(c)
			{
			case 'i':			/* request id */
				opt_i = optarg;
				break;
			case 'c':			/* copy immediately -- PPR does this anyway */
				uprint_set_immediate_copy(upr, TRUE);
				break;
			case 'd':			/* destination */
				uprint_set_dest(upr, optarg);
				break;
			case 'f':			/* form name */
				uprint_set_form(upr, optarg);
				break;
			case 'H':			/* special handling */
				opt_H = optarg;
				break;
			case 'm':			/* send mail on normal completion */
				uprint_set_notify_email(upr, TRUE);
				break;
			case 'n':			/* number of copies */
				uprint_set_copies(upr, atoi(optarg));
				break;
			case 'o':			/* printer options */
				uprint_set_lp_interface_options(upr, optarg);
				break;
			case 'P':			/* page list */
				uprint_set_lp_pagelist(upr, optarg);
				break;
			case 'p':			/* enable notification */

				break;
			case 'q':			/* queue priority */
				uprint_set_priority(upr, atoi(optarg));
				break;
			case 's':			/* suppress messages */
				uprint_set_show_jobid(upr, FALSE);
				break;
			case 'S':			/* character set */
				uprint_set_charset(upr, optarg);
				break;
			case 't':			/* title */
				uprint_set_jobname(upr, optarg);
				break;
			case 'T':			/* content type */
				uprint_set_content_type_lp(upr, optarg);
				break;
			case 'r':			/* content type raw */
				uprint_set_content_type_lp(upr, "-r");
				break;
			case 'w':			/* write a message at job completion */
				uprint_set_notify_write(upr, TRUE);
				break;
			case 'y':			/* mode list */
				uprint_set_lp_filter_modes(upr, optarg);
				break;
			default:			/* ignore illegal options as */
				break;			/* the real lp does */
			}
		}

	if(opt_i || opt_H)
		{
		fprintf(stderr, _("%s: lp -i and -H are not yet implemented\n"), myname);
		uprint_delete(upr);
		return 1;
		}

	/* If the print destination has not yet been
	   determined, determine it now. */
	if(!uprint_get_dest(upr))
		uprint_set_dest(upr, uprint_default_destinations_lp());

	/* If there are file names, tell uprint about them: */
	if(argc > optind)
		uprint_set_files(upr, (const char **)&argv[optind]);

	/* Look for a PPR responder we can use. */
	{
	char *p;
	if((p = getenv("PPR_RESPONDER")))
		uprint_set_ppr_responder(upr, p);
	if((p = getenv("PPR_RESPONDER_ADDRESS")))
		uprint_set_ppr_responder_address(upr, p);
	if((p = getenv("PPR_RESPONDER_OPTIONS")))
		uprint_set_ppr_responder_options(upr, p);
	}

	/* Print the job. */
	{
	int return_code;

	if((return_code = uprint_print(upr, TRUE)) >= 0)
		{
		/* Child ran, nothing more to do. */
		}
	/* Unclaimed queue: */
	else if(uprint_errno == UPE_UNDEST)
		{
		fprintf(stderr, _("%s: Print queue \"%s\" not found.\n"), myname, uprint_get_dest(upr));
		/* This is the code that System V lp uses for unknown queue: */
		return_code = 1;
		}
	/* If queue found but executing program failed, */
	else
		{
		/* An arbitrary value intended to indicated a major failure: */
		return_code = 255;
		}

	/* We are done with the UPRINT structure: */
	uprint_delete(upr);

	return return_code;
	}
	} /* end of main() */

/* end of file */

